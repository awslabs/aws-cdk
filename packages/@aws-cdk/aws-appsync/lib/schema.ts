import { IBucket } from '@aws-cdk/aws-s3';
import { CfnGraphQLSchema } from './appsync.generated';
import { GraphqlApi } from './graphqlapi';
import { AssetSchema, CodeSchema, S3Schema } from './private/schema-implementation';
import { IIntermediateType } from './schema-base';
import { ResolvableField } from './schema-field';
import { ObjectType } from './schema-intermediate';

/**
 * The Schema for a GraphQL Api
 *
 * If no options are configured, schema will be generated
 * code-first.
 */
export abstract class Schema {
  /**
   * Generate a Schema from file
   *
   * @returns `Schema` with immutable schema defintion
   * @param filePath the file path of the schema file
   */
  public static fromAsset(filePath: string): Schema {
    return new AssetSchema(filePath);
  }

  /**
   * Point GraphQL Schema to a S3 Location
   *
   * @returns `Schema` with immutable schema defintion
   * @param bucket the s3 bucket that holds the GraphQL Schema
   * @param key the object key for the schema
   */
  public static fromBucket(bucket: IBucket, key: string): Schema {
    return new S3Schema(bucket, key);
  }

  /**
   * Generate a Schema that can be edited through a code-first
   * approach
   *
   * @returns `Schema` with mutable schema defintion
   */
  public static fromCode(): Schema {
    return new CodeSchema();
  }

  /**
   * The definition for this schema
   */
  public readonly abstract definition: string;

  /**
   * Called when the GraphQL Api is initialized to allow this object to bind
   * to the stack.
   */
  public abstract bind(api: GraphqlApi): CfnGraphQLSchema;

  /**
   * Escape hatch to add to Schema as desired. Will always result
   * in a newline.
   */
  public abstract addToSchema(addition: string, delimiter?: string): void;
  /**
   * Add a query field to the schema's Query. CDK will create an
   * Object Type called 'Query'. For example,
   *
   * type Query {
   *   fieldName: Field.returnType
   * }
   */
  public abstract addQuery(fieldName: string, field: ResolvableField): ObjectType;

  /**
   * Add a mutation field to the schema's Mutation. CDK will create an
   * Object Type called 'Mutation'. For example,
   *
   * type Mutation {
   *   fieldName: Field.returnType
   * }
   */
  public abstract addMutation(_fieldName: string, _field: ResolvableField): ObjectType;

  /**
   * Add a subscription field to the schema's Subscription. CDK will create an
   * Object Type called 'Subscription'. For example,
   *
   * type Subscription {
   *   fieldName: Field.returnType
   * }
   */
  public abstract addSubscription(_fieldName: string, _field: ResolvableField): ObjectType;

  /**
   * Add type to the schema
   */
  public abstract addType(_type: IIntermediateType): IIntermediateType;
}