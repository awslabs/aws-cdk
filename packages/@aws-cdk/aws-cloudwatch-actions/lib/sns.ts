import * as cloudwatch from '@aws-cdk/aws-cloudwatch';
import { ServicePrincipal } from '@aws-cdk/aws-iam';
import * as sns from '@aws-cdk/aws-sns';
import { Construct } from '@aws-cdk/core';

/**
 * Use an SNS topic as an alarm action
 */
export class SnsAction implements cloudwatch.IAlarmAction {
  constructor(private readonly topic: sns.ITopic) {
  }

  /**
   * Returns an alarm action configuration to use an SNS topic as an alarm action
   */
  public bind(_scope: Construct, _alarm: cloudwatch.IAlarm): cloudwatch.AlarmActionConfig {
    if (this.topic.masterKey) {
      this.topic.masterKey.grantEncryptDecrypt(new ServicePrincipal('cloudwatch.amazonaws.com'));
    }
    return { alarmActionArn: this.topic.topicArn };
  }
}
