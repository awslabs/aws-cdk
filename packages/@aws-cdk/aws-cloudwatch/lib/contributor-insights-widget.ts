import * as cdk from '@aws-cdk/core';
import { LegendPosition } from './graph';
import { ConcreteWidget } from './widget';
/**
 * The number of top contributors to show.
 */
export enum TopContributors {
  /**
   * The top 10 contributors
   */
  TOP10 = 10,
  /**
   * The top 25 contributors
   */
  TOP25 = 25,
  /**
   * The top 50 contributors
   */
  TOP50 = 50,
  /**
   * The top 100 contributors
   */
  TOP100 = 100,
}

/**
 * Statistic to use over the aggregation period
 */
export enum OrderStatistic {
  /**
   * All values submitted for the matching metric added together.
   * This statistic can be useful for determining the total volume of a metric.
   */
  SUM = 'Sum',
  /**
   * The highest value observed during the specified period.
   * You can use this value to determine high volumes of activity for your application.
   */
  MAXIMUM = 'Maximum'
}

/**
 * Properties for a Contributor Insights widget
 */
export interface ContributorInsightsWidgetProps {
  /**
   * Title for the widget
   *
   * @default No title
   */
  readonly title?: string;

  /**
   * Insights rule to display
   */
  readonly ruleName: string;

  /**
   * The selection of contributors to display
   *
   * @default Top 10
   */
  readonly contributorSelection?: TopContributors;

  /**
   * Position of the legend
   *
   * @default - bottom
   */
  readonly legendPosition?: LegendPosition;

  /**
   * The region the metrics of this widget should be taken from
   *
   * @default Current region
   */
  readonly region?: string;

  /**
   * Whether the graph should be shown as stacked lines
   *
   * @default false
   */
  readonly stacked?: boolean;

  /**
   * What function to use for aggregating.
   *
   * @default Statistic.SUM
   */
  readonly statistic?: OrderStatistic;

  /**
   * The period over which the statistics are applied.
   *
   * @default Duration.minutes(5)
   */
  readonly period?: cdk.Duration;

  /**
   * Width of the widget, in a grid of 24 units wide
   *
   * @default 6
   */
  readonly width?: number;

  /**
   * Height of the widget
   *
   * @default 6
   */
  readonly height?: number;

  /**
   * Account the contributor insights data comes from
   *
   * @default Deployment account
   */
  readonly account?: string;
}

/**
 * Display contributor insights
 */
export class ContributorInsightsWidget extends ConcreteWidget {
  private readonly props: ContributorInsightsWidgetProps;
  private period: number;

  constructor(props: ContributorInsightsWidgetProps) {
    super(props.width || 6, props.height || 6);

    this.period = (props.period || cdk.Duration.minutes(5)).toSeconds();
    if (![60, 300, 900, 3600, 21600, 86400].includes(this.period)) {
      throw new Error(
        `'period' must be 60s, 300s, 900s, 3600s, 21600s, or 86400s, received ${this.period}`,
      );
    }
    this.props = props;
  }

  public toJson(): any[] {
    return [
      {
        type: 'metric',
        width: this.width,
        height: this.height,
        x: this.x,
        y: this.y,
        stacked: this.props.stacked || false,
        properties: {
          accountId: this.props.account,
          period: this.period,
          view: 'timeSeries',
          title: this.props.title,
          region: this.props.region || cdk.Aws.REGION,
          insightRule: {
            maxContributorCount:
              this.props.contributorSelection || TopContributors.TOP10,
            orderBy: this.props.statistic || OrderStatistic.SUM,
            ruleName: this.props.ruleName,
          },
          legend:
            this.props.legendPosition !== undefined
              ? { position: this.props.legendPosition }
              : undefined,
        },
      },
    ];
  }
}
