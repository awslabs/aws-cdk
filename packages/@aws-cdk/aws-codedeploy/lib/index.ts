export * from './rollback-config';
export * from './ecs';
export * from './lambda';
export * from './server';
export * from './trigger-configuration';

// AWS::CodeDeploy CloudFormation Resources:
export * from './codedeploy.generated';
