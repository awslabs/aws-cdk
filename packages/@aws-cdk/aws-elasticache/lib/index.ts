// AWS::ElastiCache CloudFormation Resources:
export * from './elasticache.generated';
export * from './parameter-group';
