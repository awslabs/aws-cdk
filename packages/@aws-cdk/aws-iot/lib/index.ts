export * from './topic-rule';
export * from './topic-rule-action';
// AWS::IoT CloudFormation Resources:
export * from './iot.generated';
