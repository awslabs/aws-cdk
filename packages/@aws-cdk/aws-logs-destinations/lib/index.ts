export * from './lambda';
export * from './kinesis';
export * from './kinesisfirehose';
