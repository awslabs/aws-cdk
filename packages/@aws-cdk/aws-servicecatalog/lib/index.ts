export * from './common';
export * from './constraints';
export * from './cloudformation-template';
export * from './portfolio';
export * from './product';
export * from './tagoptions';

// AWS::ServiceCatalog CloudFormation Resources:
export * from './servicecatalog.generated';
