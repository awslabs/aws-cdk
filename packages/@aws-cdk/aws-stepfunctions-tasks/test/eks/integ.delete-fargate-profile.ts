import * as sfn from '@aws-cdk/aws-stepfunctions';
import * as cdk from '@aws-cdk/core';
import { EksDeleteFargateProfile, EksCreateFargateProfile } from '../../lib';

/*
 * Stack verification steps:
 * * aws stepfunctions start-execution --state-machine-arn <deployed state machine arn> : should return execution arn
 * * aws stepfunctions describe-execution --execution-arn <exection-arn generated before> : should return status as SUCCEEDED and a query-execution-id
 * * aws eks describe-fargate-profile --cluster-name <value> --fargate-profile-name <value> : should return status of Fargate Profile
 * * aws eks list-fargate-profile --cluster-name <value> --fargate-profile-name <value> : should not return Fargate Profile
 */
const app = new cdk.App();
const stack = new cdk.Stack(app, 'aws-stepfunctions-tasks-eks-delete-fargate-profile-integ');

const fargateProfileSelector = {
  Namespace: 'namespace',
  Labels: { 'my-label': 'my-value' },
};

const createFargateProfileJob = new EksCreateFargateProfile(stack, 'Create a Fargate Profile', {
  fargateProfileName: 'fargateprofilename',
  clusterName: 'clustername',
  podExecutionRole: '*',
  selectors: [fargateProfileSelector],
});

const deleteFargateProfileJob = new EksDeleteFargateProfile(stack, 'Delete a Fargate Profile', {
  clusterName: 'clustername',
  fargateProfileName: 'fargateprofilename',
});


const chain = sfn.Chain.start(createFargateProfileJob).next(deleteFargateProfileJob);

const sm = new sfn.StateMachine(stack, 'StateMachine', {
  definition: chain,
  timeout: cdk.Duration.seconds(30),
});

new cdk.CfnOutput(stack, 'stateMachineArn', {
  value: sm.stateMachineArn,
});


app.synth();
