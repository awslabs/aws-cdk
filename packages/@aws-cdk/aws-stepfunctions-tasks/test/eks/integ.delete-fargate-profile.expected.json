{
  "Resources": {
    "StateMachineRoleB840431D": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": {
                  "Fn::Join": [
                    "",
                    [
                      "states.",
                      {
                        "Ref": "AWS::Region"
                      },
                      ".amazonaws.com"
                    ]
                  ]
                }
              }
            }
          ],
          "Version": "2012-10-17"
        }
      }
    },
    "StateMachineRoleDefaultPolicyDF1E6607": {
      "Type": "AWS::IAM::Policy",
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "eks:CreateFargateProfile",
              "Effect": "Allow",
              "Resource": "*"
            },
            {
              "Action": "iam:GetRole",
              "Effect": "Allow",
              "Resource": "*"
            },
            {
              "Action": "iam:PassRole",
              "Condition": {
                "StringEqualsIfExists": {
                  "iam:PassedToService": [
                    "eks.amazonaws.com"
                  ]
                }
              },
              "Effect": "Allow",
              "Resource": "*"
            },
            {
              "Action": "eks:DeleteFargateProfile",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition"
                    },
                    ":eks:",
                    {
                      "Ref": "AWS::Region"
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId"
                    },
                    ":fargateprofile/clustername/fargateprofilename/*"
                  ]
                ]
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "PolicyName": "StateMachineRoleDefaultPolicyDF1E6607",
        "Roles": [
          {
            "Ref": "StateMachineRoleB840431D"
          }
        ]
      }
    },
    "StateMachine2E01A3A5": {
      "Type": "AWS::StepFunctions::StateMachine",
      "Properties": {
        "RoleArn": {
          "Fn::GetAtt": [
            "StateMachineRoleB840431D",
            "Arn"
          ]
        },
        "DefinitionString": {
          "Fn::Join": [
            "",
            [
              "{\"StartAt\":\"Create a Fargate Profile\",\"States\":{\"Create a Fargate Profile\":{\"Next\":\"Delete a Fargate Profile\",\"Type\":\"Task\",\"Resource\":\"arn:",
              {
                "Ref": "AWS::Partition"
              },
              ":states:::eks:createFargateProfile\",\"Parameters\":{\"FargateProfileName\":\"fargateprofilename\",\"ClusterName\":\"clustername\",\"PodExecutionRoleArn\":\"*\",\"Selectors\":[{\"Namespace\":\"namespace\",\"Labels\":{\"my-label\":\"my-value\"}}]}},\"Delete a Fargate Profile\":{\"End\":true,\"Type\":\"Task\",\"Resource\":\"arn:",
              {
                "Ref": "AWS::Partition"
              },
              ":states:::eks:deleteFargateProfile\",\"Parameters\":{\"ClusterName\":\"clustername\",\"FargateProfileName\":\"fargateprofilename\"}}},\"TimeoutSeconds\":30}"
            ]
          ]
        }
      },
      "DependsOn": [
        "StateMachineRoleDefaultPolicyDF1E6607",
        "StateMachineRoleB840431D"
      ]
    }
  },
  "Outputs": {
    "stateMachineArn": {
      "Value": {
        "Ref": "StateMachine2E01A3A5"
      }
    }
  }
}